src/Util/Crypto.php

```diff
@@ -246,4 +247,241 @@ public static function newKeypair($bits)
+		return openssl_decrypt($data, 'aes-256-ctr', str_pad($key, 32, "\0"), OPENSSL_RAW_DATA, str_pad($iv, 16, "\0"));
+	}
+
+	/**
+	 * 
+	 * Ported from Hubzilla: https://framagit.org/hubzilla/core/blob/master/include/crypto.php
+	 * 
+	 * @param string $data
+	 * @param string $pubkey The public key.
+	 * @param string $alg    The algorithm used for encryption.
+	 * 
+	 * @return array
+	 */
+	public static function encapsulate($data, $pubkey, $alg = 'aes256cbc')
+	{
+		if ($alg === 'aes256cbc') {
+			return self::encapsulateAes($data, $pubkey);
+		}
+		return self::encapsulateOther($data, $pubkey, $alg);
+	}
+
+	/**
+	 * 
+	 * Ported from Hubzilla: https://framagit.org/hubzilla/core/blob/master/include/crypto.php
+	 * 
+	 * @param type $data
+	 * @param type $pubkey The public key.
+	 * @param type $alg    The algorithm used for encryption.
+	 * 
+	 * @return array
+	 */
+	private static function encapsulateOther($data, $pubkey, $alg)
+	{
+		if (!$pubkey) {
+			logger('no key. data: '.$data);
+		}
+		$fn = 'encrypt' . strtoupper($alg);
+		if (method_exists(__CLASS__, $fn)) {
+			// A bit hesitant to use openssl_random_pseudo_bytes() as we know
+			// it has been historically targeted by US agencies for 'weakening'.
+			// It is still arguably better than trying to come up with an
+			// alternative cryptographically secure random generator.
+			// There is little point in using the optional second arg to flag the
+			// assurance of security since it is meaningless if the source algorithms
+			// have been compromised. Also none of this matters if RSA has been
+			// compromised by state actors and evidence is mounting that this has
+			// already happened.
+			$result = ['encrypted' => true];
+			$key = openssl_random_pseudo_bytes(256);
+			$iv  = openssl_random_pseudo_bytes(256);
+			$result['data'] = base64url_encode(self::$fn($data, $key, $iv), true);
+
+			// log the offending call so we can track it down
+			if (!openssl_public_encrypt($key, $k, $pubkey)) {
+				$x = debug_backtrace();
+				logger('RSA failed. ' . print_r($x[0], true));
+			}
+
+			$result['alg'] = $alg;
+			$result['key'] = base64url_encode($k, true);
+			openssl_public_encrypt($iv, $i, $pubkey);
+			$result['iv'] = base64url_encode($i, true);
+
+			return $result;
+		} else {
+			$x = ['data' => $data, 'pubkey' => $pubkey, 'alg' => $alg, 'result' => $data];
+			Addon::callHooks('other_encapsulate', $x);
+
+			return $x['result'];
+		}
+	}
+
+	/**
+	 * 
+	 * Ported from Hubzilla: https://framagit.org/hubzilla/core/blob/master/include/crypto.php
+	 * 
+	 * @param string $data
+	 * @param string $pubkey
+	 * 
+	 * @return array
+	 */
+	private static function encapsulateAes($data, $pubkey)
+	{
+		if (!$pubkey) {
+			logger('aes_encapsulate: no key. data: ' . $data);
+		}
+
+		$key = openssl_random_pseudo_bytes(32);
+		$iv  = openssl_random_pseudo_bytes(16);
+		$result = ['encrypted' => true];
+		$result['data'] = base64url_encode(AES256CBC_encrypt($data, $key, $iv), true);
```

Comment:
This function doesn't exist in the Friendica codebase. I belive it should be `self:: encryptAES256CBC`.